evaluationquestions.txt

```diff
@@ -0,0 +1,91 @@
+what is the difference between Synchronous and ASynchronous?
+Synchronous calls refer to a callback where the code execution waits for an event before continuing, while Asynchronous calls, on the other hand, refer to a callback that does not block the execution of the program.
+
+what are promises ?
+Promis is program commiting to do something if resolved, it represent processes that are already happening, which can be chained with callback functions.
+
+write a promise that resolves when getting a value.
+Promise.resolve('resolved').then(
+    (result) => { 
+       console.log(result);
+    },
+    (error) => { 
+       console.log(error);
+    }
+  );
+
+write a promise that reject when getting an error.
+Promise.reject(new Error('fail')).then(
+    (result) => { 
+       console.log(result);
+    },
+    (error) => { 
+       console.log(error);
+    }
+  );
+
+should we use syn await, or .then().
+await is just an internal version of .then() (doing basically the same thing), where the reason to choose one over the other doesn't really have to do with performance, but has to do with desired coding style or coding convenience.
+If we have multiple successive Asynchronous calls it is better to use await, rather than bracketing them each inside a .then() handler to chain them properly, you could just use await and have simpler looking code
+
+what is promise.all
+Promise.all([promises]) accepts a collection (for example, an array) of promises as an argument and executes them in parallel.
+This method waits for all the promises to resolve and returns the array of promise results. If any of the promises reject or execute to fail due to an error, all other promise results will be ignored.
+
+what is promise.race
+Promise.race([promises]) – It waits for the first (quickest) promise to settle, and returns the result/error accordingly.
+
+when do we use promise.all and when do we use promise.race
+Both of the Promise methods receive an array of promises, however, you’ll want to choose one over the other depending on what you need to accomplish.
+Promise.all will attempt to fulfill all of them. Exits early if just 1 promise gets rejected.
+Promise.race will returns the first promise that is settled. A settled promise can either be resolved or rejected.
+
+is setTimeout an ASynchronous function
+setTimeout()is an asynchronous function, meaning that the timer function will not pause execution of other functions in the functions stack
+
+what are callbacks
+passed function is referred as callback function, mean function that can then be called inside another functions
+There are two types of callbacks. A callback passed to a function f can be invoked synchronously before f returns, or it can be deferred to execute asynchronously some time after f returns
+
+////////////////    part C      ////////////////
+using lodash function, get the employee with the highest age.
+let highestage =_(employees).map('age').value();
```

Comment:
could you please explain what you did here?

Reply:
here i get the array of age, then bring the maximum from this array